### PR TITLE
Flow with evaporation

### PR DESCRIPTION
**Pull request by mschreter**
This PR couples the two-phase flow problem with evaporation, according to 

1. modification of the interface velocity, implemented in PR #76
2. modification of the balance of mass by adding the right-handside term
<p align="center">
 
</p>

### CODE REVIEW

**Comment by mschreter** [comment]
All tests are passing locally, so this PR would be ready from my side :).

File: include/meltpooldg/flow/adaflo_wrapper.hpp [hunk 1]
```diff
@@ -68,20 +76,23 @@ namespace MeltPoolDG::Flow
...
+      this->dof_index_hanging_nodes_u =
+        scratch_data.attach_constraint_matrix(navier_stokes.get_hanging_node_constraints_u());
 
       const auto &adaflo_params = base_in->parameters.adaflo_params.get_parameters();
 
       this->quad_index_u =
         adaflo_params.use_simplex_mesh ?
           scratch_data.attach_quadrature(Simplex::QGauss<dim>(adaflo_params.velocity_degree + 1)) :
-          scratch_data.attach_quadrature(QGauss<dim>(adaflo_params.velocity_degree + 1));
+          scratch_data.attach_quadrature(QGauss<1>(adaflo_params.velocity_degree + 1));
```
Comment (peterrum) [inline include/meltpooldg/flow/adaflo_wrapper.hpp:91]: Why did you make this change? `MatrixFree` will internally implicitly convert it to `QGauss<dim>`? But no need change it back!

File: include/meltpooldg/flow/adaflo_wrapper.hpp [hunk 2]
```diff
@@ -250,6 +303,17 @@ namespace MeltPoolDG::Flow
       vectors.push_back(&navier_stokes.solution_old.block(1));
     }
 
+    void
+    distribute_constraints() override
+    {
+      navier_stokes.get_constraints_u().distribute(navier_stokes.solution.block(0));
+      navier_stokes.get_hanging_node_constraints_u().distribute(
+        navier_stokes.solution_old.block(0));
```
Comment (peterrum) [inline include/meltpooldg/flow/adaflo_wrapper.hpp:311]: Isn't ` navier_stokes.get_hanging_node_constraints_u()` part of `navier_stokes.get_constraints_u()`? I thought both are merged at some time?

File: include/meltpooldg/advection_diffusion/advection_diffusion_operator.hpp [hunk 3]
```diff
@@ -281,25 +285,6 @@ namespace MeltPoolDG
           false); // rhs should not be zeroed out in order to consider inhomogeneous dirichlet BC
       }
 
-    private:
-      static vector
-      map_to_vector(const scalar &in)
```
Comment (peterrum) [inline include/meltpooldg/advection_diffusion/advection_diffusion_operator.hpp:286]: Oh, it looks like the function was already called this way!? Did I call it this way!? I remember I wrote these a long time ago to make 1D running...

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 4]
```diff
@@ -52,77 +52,167 @@ namespace MeltPoolDG::Evaporation
...
+                {
+                  const auto n_phi =
+                    MeltPoolDG::VectorTools::normalize<dim>(normal_vec.get_value(q_index));
+
+                  // the factor of 0.5 is needed to ensure that the integral of phi * n * 0.5
+                  // over the volume is equal to 1 and represents the approximation of a delta
+                  // function. @todo -- better solution?
+                  mass_flux.submit_value(-(1. / evaporation_data.density_liquid -
+                                           1. / evaporation_data.density_gas) *
+                                           evaporation_data.evaporative_mass_flux *
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:200]: How can `evaporative_mass_flux` be a parameter?

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 5]
```diff
@@ -52,77 +52,167 @@ namespace MeltPoolDG::Evaporation
...
+                                           heaviside.get_gradient(q_index) * n_phi * 0.5,
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:201]: I guess this line is the approximation of the delta function? In contrast to the force computation, we multiply it with the normal vector to get a scalar, don't we?

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 6]
```diff
@@ -52,77 +52,167 @@ namespace MeltPoolDG::Evaporation
...
+
+              for (unsigned int q_index = 0; q_index < mass_flux.n_q_points; ++q_index)
+                {
+                  const auto n_phi =
+                    MeltPoolDG::VectorTools::normalize<dim>(normal_vec.get_value(q_index));
+
+                  // the factor of 0.5 is needed to ensure that the integral of phi * n * 0.5
+                  // over the volume is equal to 1 and represents the approximation of a delta
+                  // function. @todo -- better solution?
+                  mass_flux.submit_value(-(1. / evaporation_data.density_liquid -
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:198]: What is the reason for the double negation here?

File: include/meltpooldg/evaporation/evaporation_operation.hpp [hunk 7]
```diff
@@ -52,77 +52,167 @@ namespace MeltPoolDG::Evaporation
...
+              mass_flux.reinit(cell);
+
+              for (unsigned int q_index = 0; q_index < mass_flux.n_q_points; ++q_index)
+                {
+                  const auto n_phi =
+                    MeltPoolDG::VectorTools::normalize<dim>(normal_vec.get_value(q_index));
+
+                  // the factor of 0.5 is needed to ensure that the integral of phi * n * 0.5
+                  // over the volume is equal to 1 and represents the approximation of a delta
+                  // function. @todo -- better solution?
```
Comment (peterrum) [inline include/meltpooldg/evaporation/evaporation_operation.hpp:197]: Maybe we can try out in the next week(s) some sharp-interface methods ;)

**Review by peterrum — APPROVED** [review]
Nothing serious from my side!

File: tests/stefans_problem_with_flow_amr.with_adaflo_support=on.mpirun=4.output [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+-- Statistics --                    min      avg      max avg/call  p_min  p_max
+-- Statistics -- memory [MB] :      306      307      308           1      0     
+-- Statistics -- nln solver  :    0.519    0.519    0.519    0.519  0      3     
+-- Statistics --  lin solver :    0.349    0.349    0.349    0.116  0      3     
+-- Statistics --   mat-vec   :    0.232    0.232    0.232   0.0043  2      1     
+-- Statistics --   full prec :   0.0854   0.0857   0.0859  0.00159  1      2     
+-- Statistics --   velocity  :  0.00418   0.0044  0.00459 8.14e-05  1      2     
+-- Statistics --   div matrix:   0.0698   0.0699     0.07  0.00129  1      3     
+-- Statistics --   pres mass : 0.000656 0.000708 0.000749 1.31e-05  3      1     
+-- Statistics --   pres Poiss:   0.0103   0.0104   0.0105 0.000193  3      0
```
[the remark this answers is not in the document]
Comment (mschreter) [reply tests/stefans_problem_with_flow_amr.with_adaflo_support=on.mpirun=4.output:33]: True, thank you.

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 4 above)
Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp:200]: This is only intermediate as long as the solution of the thermal problem is omitted.

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 5 above)
Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp:201]: The term `heaviside.get_gradient(q_index) * n_phi * 0.5` is the approximation of the delta function. I am not very happy with the solution yet.

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 6 above)
Comment (mschreter) [reply include/meltpooldg/evaporation/evaporation_operation.hpp:198]: First, I did not have the negative sign and the results did not look correct ;-). So, I've introduced the negative sign. I'll rewrite that.

File: include/meltpooldg/flow/adaflo_wrapper.hpp (hunk 2 above)
Comment (mschreter) [reply include/meltpooldg/flow/adaflo_wrapper.hpp:311]: This is true. We can discuss this point later.

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 6 above)
Comment (peterrum) [reply include/meltpooldg/evaporation/evaporation_operation.hpp]: Maybe the reason that adaflo expects the "residual"?

**Review by peterrum — APPROVED** [review]
Feel free to merge!

File: include/meltpooldg/evaporation/evaporation_operation.hpp (hunk 7 above)
Comment (peterrum) [reply include/meltpooldg/evaporation/evaporation_operation.hpp:197]: Similarly to https://github.com/kronbichler/adaflo/pull/30, the code might look like:
```cpp
for (unsigned int q = 0; q < n_points; ++q)
  phi_rhs.submit_value(phi_mass.get_value(q) * JxW[q], q);
```
or with `phi_temperatore.get_gradient(q)`.

I think a code like this is worth a try, @mschreter!

**Pull request merged**